src/metabase/driver/postgres.clj

```diff
@@ -279,6 +276,56 @@
   (and (str/starts-with? database-type "\"")
        (str/ends-with? database-type "\"")))
 
+(defmethod sql.qp/->honeysql [:postgres :convert-timezone]
+  [driver [_ arg target-timezone source-timezone]]
+  (let [expr         (sql.qp/->honeysql driver arg)
+        timestamptz? (hx/is-of-type? expr "timestamptz")]
+    (when (and timestamptz? source-timezone)
+      (throw (ex-info "`timestamp with time zone` columns shouldn't have a `from timezone`" {:target-timezone target-timezone
+                                                                                             :source-timezone source-timezone})))
+    (let [source-timezone (or source-timezone (qp.timezone/results-timezone-id))
+          ;; If a column doesn't have a timezone, calling `timezone(source-timezone, column)` will cast it
+          ;; to `timestamp with time zone` with time zone is the `source-timezone`.
+          ;; For `timestamp with time zone` columns, Postgres will retuns all values in the report-tz regardless of
+          ;; what time zone column is in.
+          ;; So what we care is to just make sure the output of `convert-timezone` is a `timestammp with time zone`.
+          ;; And the actual conversion will be done in `format-rows` middleware.
+          expr (if-not timestamptz?
```

Comment:
I tried to explain in the comment above, but I'll reiterate here in a different way in case the comment is unclear:
- the MBQL for convert-timezone is actually not responsible for the actual timezone conversion.
- Instead the conversion is done in the `format-rows` middleware
- All this `convert-timezone` expression does in MBQL is make sure its output is `timestamp with timezone` and annotate some type info.
- The reason for this is Postgres will return a `timestamp with time zone` in the `report-tz` anyway, that means converting the timezone in MBQL is actually close to no-op.